### PR TITLE
Add connection proxy settings to Elasticsearch output

### PR DESCRIPTION
**Pull request by karolz-ms**
Implements https://github.com/Azure/diagnostics-eventflow/issues/394

### CODE REVIEW

File: src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/Configuration/ElasticSearchProxyConfiguration.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.EventFlow.Configuration
+{
+    public class ElasticSearchProxyConfiguration
+    {
+        public string Uri { get; set; }
```
Comment (xiaomi7732) [inline src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/Configuration/ElasticSearchProxyConfiguration.cs:16]: Is there a reason not to use type of `Uri` or `Uri?` directly?

File: src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/ElasticSearchOutput.cs [hunk 2]
```diff
@@ -309,6 +309,24 @@ private void Initialize(ElasticSearchOutputConfiguration esOutputConfiguration)
             {
                 connectionSettings = connectionSettings.BasicAuthentication(userName, password);
             }
+            if (!string.IsNullOrWhiteSpace(esOutputConfiguration.Proxy.Uri))
+            {
+                Uri proxyUri = null;
```
Comment (xiaomi7732) [inline src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/ElasticSearchOutput.cs:314]: How about:

```csharp
if(Uri.TryParse(esOutputConfiguration.Proxy.Uri, out Uri proxyUri)
{
    connectionSettings.Proxy(proxyUri, ..);
}
else
{
    var errorMessage = $"{nameof(ElasticSearchOutput)}: proxy URI setting value '{esOutputConfiguration.Proxy.Uri}' is not valid";
    healthReporter.ReportWarning(errorMessage, EventFlowContextIdentifiers.Configuration);
}
```

File: test/Microsoft.Diagnostics.EventFlow.Outputs.Tests/ElasticSearchOutputTests.cs [hunk 3]
```diff
@@ -43,7 +43,13 @@ public void VerifyAllConfigOptionsMappedToConfigObject()
                                     ["type"] = "date_nanos"
                                 }
                             }
-                        }
+                        },
+                        ["proxy"] = new Dictionary<string, object>
```
Comment (xiaomi7732) [inline test/Microsoft.Diagnostics.EventFlow.Outputs.Tests/ElasticSearchOutputTests.cs:47]: It is great that we could cover this case here. I am just wondering, is it common for a proxy to be provided without username and password? What do you think to add some unit tests to cover that when userName or password being null or empty?

**Review by xiaomi7732 — CHANGES_REQUESTED** [review]
Generally LGTM. Just a few questions to satisfy my curiosity. Mind give some explanations? Thanks!

File: test/Microsoft.Diagnostics.EventFlow.Outputs.Tests/ElasticSearchOutputTests.cs (hunk 3 above)
Comment (srothkin) [reply test/Microsoft.Diagnostics.EventFlow.Outputs.Tests/ElasticSearchOutputTests.cs:47]: For our on-prem environment there's no username/password for the proxy. And the proxy is plain http. The proxy is being used to limit which internet sites can be accessed from our production servers.

File: src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/ElasticSearchOutput.cs (hunk 2 above)
Comment (karolz-ms) [reply src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/ElasticSearchOutput.cs:314]: Great suggestion, will do that

File: src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/Configuration/ElasticSearchProxyConfiguration.cs (hunk 1 above)
Comment (karolz-ms) [reply src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/Configuration/ElasticSearchProxyConfiguration.cs:16]: It is a string in order to avoid dealing with malformed URIs at deserialization time.

File: test/Microsoft.Diagnostics.EventFlow.Outputs.Tests/ElasticSearchOutputTests.cs (hunk 3 above)
Comment (karolz-ms) [reply test/Microsoft.Diagnostics.EventFlow.Outputs.Tests/ElasticSearchOutputTests.cs:47]: Like @srothkin said, empty/null username and passwords are valid use cases.

Comment (xiaomi7732) [reply test/Microsoft.Diagnostics.EventFlow.Outputs.Tests/ElasticSearchOutputTests.cs:47]: @karolz-ms Your call. But what I meant was to have unit test cover when userName/password is null and the code keeps working. In other words, probably no exception when userName or password is empty.

File: src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/Configuration/ElasticSearchProxyConfiguration.cs (hunk 1 above)
Comment (xiaomi7732) [reply src/Microsoft.Diagnostics.EventFlow.Outputs.ElasticSearch/Configuration/ElasticSearchProxyConfiguration.cs:16]: I see. That makes sense.

**Review by xiaomi7732 — APPROVED** [review]
Sign off.
